pkg/virt-controller/watch/instancetype/upgrade_test.go

```diff
@@ -0,0 +1,257 @@
+		controller.upgrader = newMockUpgrader()
+
+		mockQueue = testutils.NewMockWorkQueue(controller.Queue)
+		controller.Queue = mockQueue
+
+		stop = make(chan struct{})
+		syncCaches(stop)
+	})
+
+	addCR := func(cr *appsv1.ControllerRevision) {
+		Expect(crInformer.GetStore().Add(cr)).To(Succeed())
+	}
+
+	addCRUpgrade := func(crUpgrade *instancetypev1beta1.ControllerRevisionUpgrade) {
+		mockQueue.ExpectAdds(1)
+		crUpgradeSource.Add(crUpgrade)
+		mockQueue.Wait()
+	}
+
+	addInProgressUpgrade := func() {
+		inprogress := instancetypev1beta1.UpgradeInProgress
+		addCRUpgrade(&instancetypev1beta1.ControllerRevisionUpgrade{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: util.NamespaceTestDefault,
+				Name:      crUpgradeName,
+			},
+			Spec: &instancetypev1beta1.ControllerRevisionUpgradeSpec{
+				TargetName: crName,
+			},
+			Status: &instancetypev1beta1.ControllerRevisionUpgradeStatus{
+				Phase: &inprogress,
+			},
+		})
+	}
+
+	addUnsetUpgrade := func() {
+		addCRUpgrade(&instancetypev1beta1.ControllerRevisionUpgrade{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: util.NamespaceTestDefault,
+				Name:      crUpgradeName,
+			},
+			Spec: &instancetypev1beta1.ControllerRevisionUpgradeSpec{
+				TargetName: crName,
+			},
+		})
+	}
+
+	expectUpgradePhase := func(phase instancetypev1beta1.ControllerRevisionUpgradePhase) {
+		client.Fake.PrependReactor("update", instancetypeapi.PluralControllerRevisionUpgradeResourceName, func(action testing.Action) (handled bool, ret runtime.Object, err error) {
+			update, ok := action.(testing.UpdateAction)
+			Expect(ok).To(BeTrue())
+
+			crUpgrade := update.GetObject().(*instancetypev1beta1.ControllerRevisionUpgrade)
+			Expect(*crUpgrade.Status.Phase).To(Equal(phase))
+
+			return true, update.GetObject(), nil
+		})
+	}
+
+	expectUpdatePhaseToInProgress := func() {
+		expectUpgradePhase(instancetypev1beta1.UpgradeInProgress)
+	}
+
+	expectUpdatePhaseToFailed := func() {
+		expectUpgradePhase(instancetypev1beta1.UpgradeFailed)
+	}
+
+	expectUpdatePhaseToSucceeded := func() {
+		expectUpgradePhase(instancetypev1beta1.UpgradeSucceeded)
+	}
+
+	assertExecutedWithoutReenqueue := func() {
+		Expect(mockQueue.Len()).To(Equal(1))
+		Expect(controller.Execute()).To(BeTrue())
+		Expect(mockQueue.Len()).To(Equal(0))
+		Expect(mockQueue.GetRateLimitedEnqueueCount()).To(Equal(0))
+	}
+
+	assertExecutedWithReenqueue := func() {
+		Expect(mockQueue.Len()).To(Equal(1))
+		Expect(controller.Execute()).To(BeTrue())
+		Expect(mockQueue.Len()).To(Equal(0))
+		Expect(mockQueue.GetRateLimitedEnqueueCount()).To(Equal(1))
+	}
+
+	Context("ControllerRevisionUpgrade", func() {
+		BeforeEach(func() {
+			addCR(&appsv1.ControllerRevision{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: util.NamespaceTestDefault,
+					Name:      crName,
+				},
+			})
+		})
+
+		It("should reenqueue on failure to find ControllerRevisionUpgrade", func() {
+			mockQueue.Add("non-existing-crUpgrade-key")
+			assertExecutedWithReenqueue()
```

Comment:
Will it retry indefinitely?

Reply:
Yeah as above, I don't mind reworking this to fail and forget.